src/Kestrel.Core/Internal/ConnectionHandler.cs

```diff
@@ -24,38 +28,73 @@ public ConnectionHandler(ListenOptions listenOptions, ServiceContext serviceCont
             _application = application;
         }
 
-        public IConnectionContext OnConnection(IConnectionInformation connectionInfo)
+        public void OnConnection(IFeatureCollection features)
         {
-            var inputPipe = connectionInfo.PipeFactory.Create(GetInputPipeOptions(connectionInfo.InputWriterScheduler));
-            var outputPipe = connectionInfo.PipeFactory.Create(GetOutputPipeOptions(connectionInfo.OutputReaderScheduler));
+            var connectionContext = new DefaultConnectionContext(features);
+
+            var transportFeature = connectionContext.Features.Get<IConnectionTransportFeature>();
+
+            var inputPipe = transportFeature.PipeFactory.Create(GetInputPipeOptions(transportFeature.InputWriterScheduler));
+            var outputPipe = transportFeature.PipeFactory.Create(GetOutputPipeOptions(transportFeature.OutputReaderScheduler));
 
             var connectionId = CorrelationIdGenerator.GetNextId();
             var frameConnectionId = Interlocked.Increment(ref _lastFrameConnectionId);
 
+            // Set the transport and connection id
+            connectionContext.ConnectionId = connectionId;
+            transportFeature.Connection = new PipeConnection(inputPipe.Reader, outputPipe.Writer);
+            var applicationConnection = new PipeConnection(outputPipe.Reader, inputPipe.Writer);
```

Comment:
No, the application reads from and writes to the transport thats why the application gets a handle to the transport.

The transport reads from and writes to the applications, thats why the transport gets a handle to the application.